libkbfs/rekey_queue.go

```diff
@@ -37,148 +60,114 @@ var _ RekeyQueue = (*RekeyQueueStandard)(nil)
 func NewRekeyQueueStandard(config Config) *RekeyQueueStandard {
 	log := config.MakeLogger("RQ")
 	rkq := &RekeyQueueStandard{
-		config: config,
-		log:    log,
+		config:   config,
+		log:      log,
+		queue:    make(chan rekeyQueueEntry, rekeyQueueSize),
+		pendings: make(map[tlf.ID]<-chan error),
 	}
 	return rkq
 }
 
+func (rkq *RekeyQueueStandard) doneLocked(entry rekeyQueueEntry, err error) {
+	entry.ch <- err
+	close(entry.ch)
+	delete(rkq.pendings, entry.id)
+	rkq.wg.Done()
+}
+
+func (rkq *RekeyQueueStandard) done(entry rekeyQueueEntry, err error) {
+	rkq.lock.Lock()
+	defer rkq.lock.Unlock()
+	rkq.doneLocked(entry, err)
+}
+
+func (rkq *RekeyQueueStandard) work(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case entry := <-rkq.queue:
+			newCtx := ctxWithRandomIDReplayable(ctx, CtxRekeyIDKey,
+				CtxRekeyOpID, nil)
+			rkq.log.CDebugf(newCtx, "Processing rekey for %s", entry.id)
+			err := rkq.config.KBFSOps().Rekey(newCtx, entry.id)
+			rkq.done(entry, err)
+		}
+	}
+}
+
+func (rkq *RekeyQueueStandard) ensureRunningLocked() {
+	if rkq.cancel != nil {
+		return
+	}
+	var ctx context.Context
+	ctx, rkq.cancel = context.WithCancel(context.Background())
+	for i := 0; i < numRekeyWorkers; i++ {
+		go rkq.work(ctx)
+	}
+}
+
 // Enqueue implements the RekeyQueue interface for RekeyQueueStandard.
 func (rkq *RekeyQueueStandard) Enqueue(id tlf.ID) <-chan error {
 	rkq.log.Debug("Enqueueing %s for rekey", id)
-	c := make(chan error, 1)
-	rkq.wg.Add(1)
-	func() {
-		rkq.queueMu.Lock()
-		defer rkq.queueMu.Unlock()
-		if rkq.cancel == nil {
-			// create a new channel
-			rkq.hasWorkCh = make(chan struct{}, 1)
-			// spawn goroutine if needed
-			var ctx context.Context
-			ctx, rkq.cancel = context.WithCancel(context.Background())
-			go rkq.processRekeys(ctx, rkq.hasWorkCh)
-		}
-		rkq.queue = append(rkq.queue, rekeyQueueEntry{id, c})
-	}()
-	// poke the channel
-	select {
-	case rkq.hasWorkCh <- struct{}{}:
-	default:
+
+	if ch := rkq.GetRekeyChannel(id); ch != nil {
+		return ch
 	}
-	return c
+
+	rkq.wg.Add(1)
+
+	rkq.lock.Lock()
+	defer rkq.lock.Unlock()
+
+	rkq.ensureRunningLocked()
+
+	ch := make(chan error, 1)
+	rkq.queue <- rekeyQueueEntry{id: id, ch: ch}
```

Comment:
Appending to the queue while holding the lock is probably not a good idea, since it can block.

In fact, `folderBranchOps` calls `Enqueue` while holding an important lock as well, so if this blocks it might cause issues.

I understand why it's attractive to put the ids in a channel, but maybe we should do something simple like this to be safe:

```go
select {
  case rkq.queue <- rekeyQueueEntry{id: id, ch: ch}:
  default:
    // The queue is full; avoid blocking by spawning a goroutine.
    rkq.log.CDebugf(context.TODO(), "Rekey queue is full; enqueuing %s in the background", id)
    go func() { rkq.queue <- rekeyQueueEntry{id: id, ch: ch} }
}

```